base/serialize.jl

```diff
@@ -420,11 +420,15 @@ function serialize_any(s::SerializationState, x)
     else
         t.mutable && serialize_cycle(s, x) && return
         serialize_type(s, t)
-        for i in 1:nf
-            if isdefined(x, i)
-                serialize(s, getfield(x, i))
-            else
-                writetag(s.io, UNDEFREF_TAG)
+        if t.pointerfree && ccall(:jl_datatype_haspadding, Cint, (Any,), t)==0
+            write(s.io, pointer_from_objref(x), sizeof(x))
```

Comment:
the pointer returned here may be to a box that is not gc-rooted. in general, `pointer_from_objref` should never be called on something that is `immutable` and `pointerfree` (I think we should make it an error)

i'm looking into possible alternatives that would provide the same benefit, but doesn't require calling pointer_from_objref